### ISSUE TITLE
Python 2->3 Upgrade

### ISSUE THREAD

**Issue by geordie666** [desihub/desitarget#60]
Need to rewrite for python 3.5 upgrade. ADM will do this except for the mocks.

**Comment by sbailey** [comment]
2to3 changes are done already.  I pushed a new `py3` branch with an update so that`isinstance` will compare to `numbers.Integral` instead of `int`, since `isinstance(np.int64(1), int)` is False under py3.

I also updated `test_readwrite_tractor` to use fitsio for the final check (for bytes vs. unicode reasons...), but that fails when comparing the 'TYPE' column due to the issue in esheldon/fitsio#99 : cfitsio pads strings in binary tables with trailing spaces such that if you write a column `data['TYPE'] == 'PSF'` and read it back in, you get `data['TYPE'] == 'PSF '` (with a trailing space).  Using `astropy.io.fits` instead fails for different reasons: it strips the trailing spaces but tries to automatically upgrade to unicode, leaving a mismatch between bytes vs. unicode...

Although this is a cfitsio bug (and thus a problem with fitsio), we could just work around it for now such that the tests pass.

@geordie666 consider basing any future python 3 changes on this branch and push back to this branch, if you haven't already started on your latest round of updates.  You'll need the numbers.Integral stuff anyway, but we'll still need to figure out what to do about cfitsio bugs and/or bytes vs. unicode.

**Comment by geordie666** [comment]
I've been wrestling with how to get around the unicode/byte encoding. It occurs to me that because (I think) our only problematic column from the Tractor files is the "TYPE" column (which can be b'SIMP' or b'SIMP' but can equally be b'EXP ' with a trailing space or b'EXP') we could just hack that column during IO to always be padded to at least 4 characters. This code snippet does that:

for row in range(len(data)):
            data['TYPE'][row] = ('{0: <4}'.format(data['TYPE'][row].decode())).encode()

I don't think that this solution would be too slow, as it's only iterating over one column of the tractor (or sweep) files. And, it has the advantage that it should universally work however files are ingested.

I can certainly hack this into the unit test, but it might be a generally workable solution for IO.

**Comment by sbailey** [comment]
+1 for fixing this at the I/O layer instead of just hacking the unit test.

A trick I learned from Erin Sheldon: you could `numpy.char.rstrip` instead (faster, simpler):
```python
data['TYPE'] = np.char.rstrip(data['TYPE'])
```
or something more general like
```python
for colname in data.dtype.names:
    kind = data[colname].dtype.kind
    if kind == 'U' or kind == 'S':
        data[colname] = np.char.rstrip(data[colname])
```
He's working on an update to fitsio to make this an option when reading a fits file, but its not available yet.

**Comment by geordie666** [comment]
Fixed the specific ```TYPE``` problem in ```io.read_tractor``` by always padding it to a 4-string as follows:

```
if 'TYPE' in readcolumns:
     data["TYPE"] = np.char.ljust(data["TYPE"],4)
```

I need to check that doesn't break other parts of the code, but it seems to work. The ```np.char``` tip was useful, but then things read in with ```fitsio``` that doesn't pass through ```io.read_tractor``` barfs.

**Comment by sbailey** [comment]
I'm confused: isn't the whole problem that cfitsio is incorrectly padding with spaces when writing, and fitsio is leaving them there when reading?  i.e. does this `ljust` code actually do anything?  Doesn't `data["type"]` already have the irritating spaces?

And even if it doesn't, I'm not thrilled about purposefully padding with spaces, since ideally we wish that cfitsio wasn't writing spaces in the first place and an upcoming feature to fitsio will allow the user to strip them on the fly when reading in.  Note that `astropy.io.fits` and `astropy.table.Table.read` also strip the spaces whether you want them to or not.  I'd rather have `read_tractor` strip the spaces away (and perhaps provide a utility function that would strip trailing spaces from any string column from any table read with fitsio), so that we can avoid code like:
```python
psflike = (psftype == 'PSF') | (psftype == 'PSF ')
```

**Comment by sbailey** [comment]
I do have a really strong preference *not* to pad, and in fact to explicitly strip spaces so that any downstream code will also work with arrays read via astropy and/or created on the fly without having to bother with how many pad spaces to use in strings.  This type of "modification during I/O" has  precedent — e.g. in desispec we store spectra on disk as float32 but cast to float64 and ensure native endianness to facilitate downstream calculations, but cast back to float32 for writing back out.

But — if you want to just get anything in now that works for python 3 and passes tests, we can also arm wrestle over this with a beer in 2 weeks.

**Comment by geordie666** [comment]
Fair enough...I'll go back to the drawing board and do what's rigorous rather than what's expedient. It's probably better to head off potential downstream issues now.

**Comment by sbailey** [comment]
I think what I'm proposing is easy, just including this in `desitarget.io.read_tractor()`:
```python
data = fx[1].read(columns=readcolumns)
for colname in data.dtype.names:
    kind = data[colname].dtype.kind
    if kind == 'U' or kind == 'S':
        data[colname] = np.char.rstrip(data[colname])
```
And then elsewhere in the code not including extra padding spaces in literal strings when comparing to the data.  I would add this myself but I don't want to step on each other with equivalent but conflicting commits.

If you think it is more complicated than that, then one or both of us are misunderstanding something.  You could also punt to me to make the change and then I would discover any mess that I'm not realizing, but I won't edit the code until getting a green light from you.  Or give me a call to resolve in realtime.

**Comment by geordie666** [comment]
In the end I did exactly as you suggested but also added a generic whitespace stripper to io. See:

https://github.com/desihub/desitarget/commit/d0c3349ba7a4524e94ae2b52b88355d01979deeb

If people using ```fitsio.read()``` (etc.) instead call ```io.whitespace_fits_read()``` then whitespace should be stripped consistently

**Comment by sbailey** [comment]
After some discussions with @geordie666 and @tskisner, we propose to revert commit cbcd820 that merged apc_mocks into py3.  This should get us back into a cleaner state for the non-mocks portion of py3 so that it can be merged into master, while continuing with the cleanup on apc_mocks before merging it.  This comment is a heads up that I plan to do that shortly (after going to get a coffee...)

**Comment by sbailey** [comment]
Update: Abort!  For once, git saved me from disaster:
```
[py3] desitarget $ git revert cbcd820
error: Commit cbcd820645bdf984df692ce7da46fd04395c0212 is a merge but no -m option was given.
fatal: revert failed
```
Reading about the `-m` option:

> Reverting a merge commit declares that you will never want the tree
    changes brought in by the merge. As a result, later merges will
    only bring in tree changes introduced by commits that are not
    ancestors of the previously reverted merge. This may or may not be
    what you want.

Well, that's not what we want.  We will eventually want these changes just not yet.  I found a [git-scm.org blog post](https://git-scm.com/2010/03/02/undoing-merges.html) about reverting merges and then reverting the revert when you eventually want the changes again, but it seems like dangerous territory.  Re-evaluating...

**Comment by forero** [comment]
We didn't do too much work after the py3/apc_mocks merge. It might be what we want.

**Comment by sbailey** [comment]
Having abandoned the plan to back out the apc_mocks -> py3 merge, I went ahead and fixed the remaining test problems on the py3 branch and just pushed that.  Some of the problems were in the tests themselves (e.g. GRAYLAYER column is gone, replaced with the more general OBSCONDITIONS mask).  But there were a few genuine bugs that are now fixed:
* mtl['OBSCONDITIONS'] wasn't being calculated
* numobs for BGS_FAINT wasn't set to 2
* typo in MTL_RESERVED_TARGETID_MIN_SKY vs. MTL_RESERVED_TARGETID_SKY

I'll open a PR for the py3 branch in a moment

**Comment by apcooper** [comment]
My work on #71 yesterday was going in the same direction as @sbailey's fixes above, but I was stymied by an `astropy.io.fits` apocalypse in [this and similar uses of writeto()](https://github.com/desihub/desitarget/blob/apc_mocks/py/desitarget/mock/mws_galaxia.py#L213) after switching to Py3 (basically a so-far inexplicable >100x jump in write time vs. Py2). Judging by the mailing list `astropy.io.fits` issues are a familiar thing? I suspect this is something specific to the format/data of the particular tables being written -- looking into it now, including an example to reproduce.

If the tests pass etc. probably can just merge apc_mocks to py3 in #72 and I'll keep working on further fixes to apc_mocks for another PR to py3 in the near future.

**Comment by apcooper** [comment]
The massive slowdown after 2->3 only occurs when the `BRICKNAME` column (the only string column) is included in the write, and also occurs when the `BRICKNAME` column is written individually.

Weirdly enough, this doesn't affect writing through Table.write(), so the obvious workaround is to use that. I didn't do that in the first place because (a) there was a marginal performance benefit in the `astropy.io.fits` route in my tests on py2.7 and (b) I ran into trouble trying to write multiple extensions for the truth table (see [here](https://github.com/desihub/desitarget/blob/apc_mocks/py/desitarget/mock/mws_wd100pc.py#L198)). Does anyone know how to do the same thing without astropy.io.fits?

**Comment by forero** [comment]
@sbailey I just ran tests on the mock routines and everything is working. From that side everything is OK.

**Comment by geordie666** [comment]
@apcooper is there a reason you can't use fitsio instead of astropy? I haven't seen a massive slowdown in Py3 writing with fitsio, although I may not be writing files as large as yours.

**Comment by apcooper** [comment]
No, there's no particular reason not to use fitsio rather than astropy.io.fits -- for some reason I had the idea that we were trying to avoid fitsio because it was a bit harder for some people to get working as a dependency (or for some other reason, I'm out of the loop there). If `fitsio` is fine (with @sbailey too?) I'll just go ahead and use that throughout. 

(General apologies for risking needless confusion with `import astropy.io.fits as fitsio` in some of those apc_mocks submodules, by the way)

**Comment by geordie666** [comment]
We're using fitsio elsewhere in desitarget, so I say go ahead and use it. Note that if you're reading in fits files, use the whitespace_fits_read wrapper in io.py (see, e.g:)

https://github.com/desihub/desitarget/commit/d0c3349ba7a4524e94ae2b52b88355d01979deeb

This will strip any white spaces from string fields upon reading in files, so we can consistently make logical comparisons between strings

**Comment by apcooper** [comment]
OK, got it, thanks. After #72  is done, I'll verify the merged code is working with my set-up and (out of necessity) switch to fitsio at the same time.

**Comment by sbailey** [comment]
A few comments:
* @apcooper are you possibly using numpy 1.10 ?  There was a huge slowdown in structured arrays with columns of strings (like BRICKNAME) that they fixed in numpy 1.11 (see numpy/numpy#6467)
* in other packages we prefer using astropy.io.fits because of trouble people have had with installing fitsio, but we've made an exception for desitarget because working with tractor files at NERSC is 30x faster with fitsio compared to astropy Tables (astropy.io.fits is somewhere in between, but also much slower than fitsio).  Those tests we done a year or two ago with py2.7 and older versions of all the packages, so it is good to report details if you see different massive slowdowns.
* but we definitely shoudn't do "import astropy.io.fits as fitsio" since they aren't API compatible :)
* Our branches have gotten a little out of hand.  I'd like to merge apc_mocks -> py3 -> master, and then have @apcooper start a new branch for future work and future PRs.  That will make it cleaner for tracking what is actually new (github seems to show the changes to the branch rather than only the differences in the source/target for a merge, so long-lived branches tend to build up huge looking PRs that obscure what is actually new).
